### ISSUE TITLE
Improvements of SCI in-abapgit

### ISSUE THREAD

**Issue by sbcgua** [larshp/abapGit#3935]
Trying to force pre-commit SCI check on one of my projects ... would probably like to do a couple of improvements

1) Make it possible to force-skip the check and go on to stage. Kind of contradicts the idea, but needed sometimes :)
2) Maybe extract an SCI button to the upper level (repo) menu as an icon ...
3) Maybe add 2 variants of SCI. One to run unit tests (only unit tests) for the whole package. Another is to run only for the staged objects (to reduce time)
4) ... consequently sci should be able to respect staged objects list
5) ... and the check should be moved to between stage and commit (is it a good idea ?)

Does the above make sense ? :)

**Comment by larshp** [comment]
1. 👍 I assume projects using pull requests, so it might not be the final version of the change
2. I worry about getting too many buttons, but guess it should only show up if configured
3: ok
4: ok
5. seems good to me, move it as early as possible

**Comment by mbtools** [comment]
This pretty much meets my requirements
- configuration per repo
- option to pick which SCI check variant to run on demand
- option to pick which SCI check variant to run automated 
- option to select whether any errors block a commit or can be skipped

"On demand" could be a single "check button" (instead of several menu items) on repo view and staging
"Automated" could be triggered at various events (before stage, before commit, scheduled)

**Comment by ed-holland** [comment]
I like the option to just run the checks on the changed objects / selected objects for stage.

to 1. : I removed the default sci variant for the repo because i do not always want to run SCI. With force skip it would be possible to define it again.

**Comment by GiantCrocodile** [comment]
What's the idea behind 1)? Right now you can define a check variant and if it's not enforced by setting `Block Commit If Code Inspection Has Errors` to true in the repository settings, you can skip it and push anyway. Is this different for forced stages?

**Comment by sbcgua** [comment]
> Right now ... you can skip it and push anyway @GiantCrocodile 

Yes, I found it too already. But in fact it means it is confusing and not obvious that you can. So at least I'd change the name to "Force stage" or "Stage anyway".

> With force skip it would be possible to define it again @ed-holland 

So... it kind of possible. 

One more idea is to skip the screen of CI if it was successful. Instead show "CI passed" green label somewhere at stage/commit level.

And more: I was imagining the process, in particular, when the check should happen. Now it is before stage. But if reduce the scope of check to staged objects - it must move to after stage. At the same time my idea is to run UTs for all and static checks for staged. So maybe run UTs before stage and statics after ? But it is really confusing!!! I think we should combine repo_view and stage pages first. (Or at least somewhen at the same time). Or move both checks after stage. Just thoughts.

**Comment by GiantCrocodile** [comment]
Hm, I don't see how this is confusing or hard to find out: The stage button is there as always and it's not grayed out or differs in any way to the other cases where you can stage. The term 'force stage' or 'stage anyway' could be misleading because you might think this is related to Git's force or that there's some issue and you wanna force anyway although there's actually no issue and you are fine to stage it anyway as it's not disabled/prevented by the configuration.

Regarding removing the screen of CI is fine for me if there's at least an indicator on the same level which shows that SCI was run indeed. This is a must have to make sure that you are aware if there was no SCI run at all due to some kind of technical issues.

Why does it have to happen after the stage if you want to reduce the object list to objects of your stage only? In our workflow it is mandatory that you can't push code with quality issues to Git so a SCI check after the stage would be too late in our scenario. Furthermore I wouldn't separate Unit Tests and ATC checks. I would leave it open to run both before staging, after staging or only one of the two before/after staging. 

Don't get me wrong: I like to see improvements regarding ATC/SCI integration but we need to plan this in detail and make sure that we don't enforce some workflow that simply doesn't fit to all projects/companies. We invested some months of time with this process and it gives us a good process with the workflow as it is; the process fits to what we wanted to realize and it has no throw backs that hinders us in any way. So I'm a bit reluctant to change it this easily.

**Comment by sbcgua** [comment]
> Force stage: I don't see how this is confusing or hard to find out

Well, it was not obvious for me initially. I always prefer to note such things. If the feature is not obvious to a person (me or the one I observe) then it may not be obvious to more. In terms of usability it would be better to emphasise overriding nature of Stage when CI didn't pass. Maybe it is color coding ? Maybe text (preferably). Any better text ideas than "Force" ? Kind of "Ignore and Stage" ??

> Regarding removing the screen of CI is fine for me if there's at least an indicator

totally. both on stage and commit pages. some kind of green/red marker "CI passed"/"CI failed"

> Why does it have to happen after the stage ... would leave it open to run both before staging

My idea was that. Unit tests should run for all. And "real" SCI check only for objects to commit. So "real" must be after stage. And UTs "as soon as possible" to give quick feedback.
In terms of usability this is crap indeed :( But ! After reading your comment I though that actually:

`Selected ≤ Changed < All`, where Selected are *usually* similar to Changed, and Changed are *usually* significantly smaller than All. Which means that limiting to changed objects is probably good enough. Which means both check can be indeed run together before the stage! Thanks for the idea !

BTW, I wonder if it is possible to create SCI variant dynamically. So that unit test variant would not need to be created manually. Could be nice to have it as a mere flag yes/no in the settings

**Comment by sbcgua** [comment]
Adding #4057 here to consolidate

**Comment by GiantCrocodile** [comment]
Hey @sbcgua,

I totally agree that there are some things that ain't that clear. For me abapGit was and is a solution that works great but has room for improvements regarding its usability. I think the Code Inspector/ATC functionality is already quite good regarding usability. Other abapGit areas might need more improvement regarding this.

>In terms of usability it would be better to emphasise overriding nature of Stage when CI didn't pass. Maybe it is color coding ? Maybe text (preferably). Any better text ideas than "Force" ? Kind of "Ignore and Stage" ??

I think working with colors isn't that good as we would need proper support for color-blind people then and it becomes quickly a mess if colors are unclear, someone doesn't know which color is used (e. g. purple vs. pink), colors get changed, etc. We should stay with a text-approach. I agree that we could emphasise the overriding nature more. `Ignore and Stage` or `Proceed anyway` would be great. The term 'Staging' isn't that clear for ppl. who are not used to this term in this context I think.

> My idea was that. Unit tests should run for all. And "real" SCI check only for objects to commit. So "real" must be after stage. And UTs "as soon as possible" to give quick feedback.

In our case this wouldn't fit: We would like to run Unit Tests and ATC (ATC can trigger Unit Tests itself) always. Performance is no issue in our case but high code quality shall and must be met anytime. So I think this must be configurable and be a decision to be decided by the quality assurance team of a company; not decided/given by abapGit. Furthermore think that could lead to problems where Unit Tests would pass but the overall scenario shouldn't pass (like single objects are coded fine and positively checked by Unit Tests but the objects are used in a non-conform way which ATC would detect as ATC checks the whole context of an object).

> Selected ≤ Changed < All, where Selected are usually similar to Changed, and Changed are usually significantly smaller than All. Which means that limiting to changed objects is probably good enough. Which means both check can be indeed run together before the stage! Thanks for the idea !

Hehe thank you :P. I like the idea to reduce the scope of objects to be checked as that highly reduces the runtime and costs of ATC checks. We would have to see how this can be done and if there's some pitfalls involved.

> BTW, I wonder if it is possible to create SCI variant dynamically. So that unit test variant would not need to be created manually. Could be nice to have it as a mere flag yes/no in the settings

Can you elaborate why an dynamic creation of the check variant is needed? I think ATC can run all Unit Tests of a specific package without manually selecting each newly created Unit Test. I've never used those Unit Test checks in ATC so I'm not sure but I think there was something like this. Creating a check variant dynamically could be an issue if you have to understand/reproduce which checks were run for specific objects, e. g. for compliance reasons.

If the CI/ATC integration gets improved, we also should improve that

1. exemptions can be requested and shown within abapGit
2. description and other information is shown as like as in the ATC result list in the ALV_GRID of the SE80

That would be important too but of course it's a different topic and should be discussed separately.

Best Regards,

**Comment by sbcgua** [comment]
- it's `Ignore and Stage` then :)

> Dynamic creation of UTs variant and Ko...

My idea was that to AG will not control if you created a variant with UTs only or with something else more. So it would be better to create UT only variant on-the fly ...

![image](https://user-images.githubusercontent.com/15635498/97635041-6124de00-1a3f-11eb-8e51-8efc4760ea44.png)


... but your doubts gave we another idea. Why limit to UTs ? You may want to run some checks (unit tests including) for the whole package always. This means having 2 SCI variants in the settings. You may specify the only the 1st, or only the 2nd, or both how you prefer! This would be flexible and lean. So my process would looks like: "always UTs for the whole package" + e.g. "abaplint for changed objects"
So thanks for another idea :)

> exemptions ...

hmm, yeah. That's a next level ... needs a separate discussion I guess.
